ordering/__init__.py

```diff
@@ -68,11 +68,13 @@ def __contains__(self, item: _T) -> bool:
         return item in self._labels
 
     def __iter__(self) -> Iterator[T]:
-        item = self._successors[self._start]
-
-        while not isinstance(item, _Sentinel):
+        successors = self._successors
+        item = successors[self._start]
+        end = self._end
+        while item is not end:
+            assert not isinstance(item, _Sentinel)
             yield item
-            item = self._successors[item]
+            item = successors[item]
```

Comment:
Does extracting the `successors` variable really make this more readable? I find the absence of `self` confusing, as it's no longer obvious that it's a member of the class.

Reply:
OrderedDict probably hides it's various attributes because they mostly wouldn't be that useful, as most of the time you'd be using an OrderedDict as a drop in replacement for a dict, thus mainly using the interface that dict provides. Plus I'm not sure I can think of any good reason why I'd actually need access to them. At this point I'd normally state that, since this is Python nothing is ever truly hidden, but as of Python 3.5 OrderedDict isn't Python -- it defaults to a C implementation (when possible) and in that case the attributes in question are hidden to the point of non-existence.

It's probably because of `slice`'s `start`, `stop`, and `step` that I wanted to use `start`/`stop`, although that's probably actually a reason against using them, as potential code involving slicing could easily get confusing.

As for `start`/`end` seeming like an odd pair, in general I'd probably use the verb pairs `start`/`stop` & `begin`/`end` and the noun pairs `start`/`finish` & `beginning`/`end`.

I can't say I'm keen on `neg_inf`/`inf`, since their labels are `0` and `1` respectively. But speaking of `inf`, I did think of `inf`/`sup` or `infimum`/`supremum` even if they aren't strictly correct. However, where they may be more correct is as class variables determining the bounds of the labels, for which I might have a use at some point... we'll see.

At first I wasn't sure about `first`/`last` since the sentinels aren't included in the values that we iterate over. However since `ordering._start in ordering` being true and `ordering._start in list(ordering)` being false irks me, I think having `first` and `last` point at the actual first and last elements (rather than the sentinels) might be the right course of action. This would unfortunately remove the current advantage that inserting terminal elements is no different from inserting intermediate elements, thus would require the introduction of new branching paths.

Also it comes with the mixed blessing that for the most part we could eliminate the need for the sentinels. The exception being when the Ordering is empty, where we would need just a single sentinel. Furthermore if we wanted we could then have public `first`/`last` properties that raise a ValueError upon an empty Ordering similar to `min([])`. On that thought another alternative to `first`/`last` not yet mentioned could be `minimum`/`maximum` or simply `min`/`max`; the symmetry of `min(ordering) == ordering.min` would be rather pleasing.

Having said that, once again it seems like the best solution comes from making Ordering follow OrderedDict, as `first`/`last` (or whatever they're named) would normally point to the OrderingItems, not the items themselves, so we can simply use `None` as the sentinel. Thus it might be best to perform any name swaps and associated changes as part of the move to the OrderedDict style.

Whilst we're talking about names if we're changing `_start`/`_end`, then we might also consider new names for `insert_start`/`insert_end`. Although with the change that the terminal attributes point to the terminal elements, we would avoid the issue that `insert_start` sounds like it would assign the new value to `_start` yet currently doesn't, so maybe `insert_first`/`insert_last` would do.

(Previously I'd have said `append` would be the obvious choice to replace `insert_end`, for consistency with lists. However that would tempt me into using `deque`'s `append`/`appendleft`, but then we have the issue if we add an extend methods, in order to be consistent with `deque.extendleft` the new items would need to be added in reverse order. Maybe the solution would be to have `append`/`prepend` and `extend`/`pre_extend`)

Ultimately I might be tempted to remove the four insert methods from Ordering in favour of using just the two of OrderingItem, if one could just write `ordering.first.insert_before(new_first)` or similar instead.

This also leads me to the question: if `first`/`last` were public should we allow assignment to them, and if so should that represent replacement or insertion? 

But, back to the subject of the insert methods: whilst not explicitly obvious, `ordering.insert_after(a, b)` is relatively easy to understand because of it's similarity to `list.insert`; it could just be me, but `ordering[a].insert_after(b)` often takes a second glance to parse correctly. Considering the (planned) OrderingItem attributes I'd suggest `insert_prev`/`insert_next`, but maybe swapping it to `after_insert`/`before_insert`, or even `pre_insert`/`post_insert` would be better?

***

Sorry this got just a *tad* long, and not always specifically on topic. For now I'll just concentrate on getting things mergeable, as is -- before moving on to figuring out how best to present an adaption to OrderedDict style without it ending up as a single monster commit. I'm thinking that starting a new branch at the initial commit might be the way to go?

Reply:
I don't see how hiding properties of OrderedDict helps when using it where you would use a dict, and it would get in the way if you want to extend OrderedDict. But I'm not too worried about this point, and we're digressing a bit.

As for `start`/`end`, the labels of `0` and `1` are entirely implementation details, while `neg_inf`/`inf` are how they are viewed from the out side, so I wouldn't discount them for that reason. `inf`/`sup` does sound kinda correct, but I'm worried that it implies they're part of the ordering. I'm not too convinced by any of the names, so I think I'll just leave it as is, for now.

You make a good point about `ordering._start` being in `ordering`. I want to avoid pointing at the actual first and last elements due to the extra special cases that occur when the ordering contains a small amount of elements. If we split the `insert_*` methods into internal/external versions, then it should be possible to make `ordering._start` not in `ordering`.

We do need some sort of insert method on Ordering, in case you want to construct an empty ordering, and then add things to it. `insert_start`/`insert_end` just seemed like the most obvious places to add to.

If you feel branching that far back is the best way to go, then feel free to do so, but I'd be surprised if that was simpler than just doing piecemeal changes to the OrderedDict style.